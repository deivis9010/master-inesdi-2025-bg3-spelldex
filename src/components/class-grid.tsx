import c from "classnames";
import classesJson from "src/data/classes.json";

import { ClassGridItem } from "./class-item";
import { useRef } from "react";

import type { CharacterClass, ClassId } from "src/models/character-class";

import styles from "./class-grid.module.css";

type Props = {
  selectedClass?: ClassId;
  background?: boolean;
  onClick: (c: ClassId | undefined) => void;
  highlight: (c: ClassId | undefined) => void;
};

const classes = classesJson as CharacterClass[];

const KEYBOARD_KEYS = [
  "ArrowRight",
  "ArrowLeft",
  "ArrowUp",
  "ArrowDown",
  "Enter",
];

export function ClassGrid({
  selectedClass,
  background,
  highlight,
  onClick,
}: Props) {
  const ref = useRef<HTMLDivElement>(null);
  const itemButtonRefs = useRef<(HTMLButtonElement | null)[]>([]);
  //bugfix: se quito el condicional para mostrar todas siempre no solo las seleccionadas
  const items =  classes;

  const currentHighlight = () => {
    //bugfix: se quito el condicional para mostrar todas siempre no solo las seleccionadas
    return  Array.from(itemButtonRefs.current).findIndex(
          (ref) => ref === document.activeElement
        );
  };

  const keyDown = (event: React.KeyboardEvent) => {
    if (!document.activeElement?.classList.contains(styles.classCell)) {
      return;
    }

    if (!KEYBOARD_KEYS.includes(event.key)) {
      return;
    }

    const currentIndex = currentHighlight();
    if (currentIndex < 0) {
      return;
    }

    let nextIndex = currentIndex;
    switch (event.key) {
      case "ArrowRight":
      case "ArrowDown":
        nextIndex = Math.min(items.length - 1, currentIndex + 1);
        break;
      case "ArrowLeft":
      case "ArrowUp":
        nextIndex = Math.max(0, currentIndex - 1);
        break;
      case "Enter":
        {
          // Navegar a la clase actualmente enfocada/highlighted
          const currentClass = items[currentIndex];
          if (currentClass) {
            onClick(currentClass.slug);
          }
          return;
        }
      default:
        break;
    }

    if (nextIndex !== currentIndex) {
      event.preventDefault();
      itemButtonRefs.current[nextIndex]?.focus();
      
      //bugfix: selecciona la clase al navegar con el teclado
      const nextClass = items[nextIndex];
      if (nextClass) {
        
        onClick(nextClass.slug);
      }

      
    }
  };

  return (
    <section
      ref={ref}
      className={c(styles.classGrid, background && styles.background)}
      aria-label="Class Grid"
      onKeyDown={keyDown}
      role="grid"
    >
      {items.map((cls, index) => (
        <ClassGridItem
          key={cls.slug}
          name={cls.name}
          classId={cls.slug}
          selected={selectedClass === cls.slug}
          highlight={highlight}
          onClick={() => onClick(cls.slug)}
          ref={(el) => {
            if (el) {
              itemButtonRefs.current[index] = el;
              if (selectedClass === cls.slug) {
                el.focus();
              }
            } else {
              itemButtonRefs.current[index] = null;
            }

            return () => {
              itemButtonRefs.current[index] = null;
            };
          }}
        />
      ))}
    </section>
  );
}
