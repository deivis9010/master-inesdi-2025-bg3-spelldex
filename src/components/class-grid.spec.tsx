import { render, screen } from "@testing-library/react";
import { ClassGrid } from "./class-grid";

describe("<ClassGrid />", () => {
  it('renders the grid: match snapshot', () => {
    const mockOnClick = jest.fn();
    const mockHighlight = jest.fn();
    
    render(<ClassGrid onClick={mockOnClick} highlight={mockHighlight} />);

    const section = screen.getByRole("grid", {
      name: "Class Grid",
    });
    expect(section).toBeInTheDocument();
    expect(section).toMatchSnapshot();
  });
});
