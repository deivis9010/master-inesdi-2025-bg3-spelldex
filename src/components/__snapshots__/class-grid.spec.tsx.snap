// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClassGrid /> renders the grid: match snapshot 1`] = `
<section
  aria-label="Class Grid"
  class=""
  role="grid"
>
  <button
    aria-label="Bard"
    aria-pressed="false"
    class=""
    tabindex="0"
  >
    <div>
      <div
        aria-hidden="true"
        class=""
      />
      <div
        aria-hidden="true"
        class=""
      />
      <div
        aria-hidden="true"
        class=""
      />
      <div
        aria-hidden="true"
        class=""
      />
      <img
        alt="Bard"
      />
    </div>
    <span>
      Bard
    </span>
  </button>
  <button
    aria-label="Cleric"
    aria-pressed="false"
    class=""
    tabindex="0"
  >
    <div>
      <div
        aria-hidden="true"
        class=""
      />
      <div
        aria-hidden="true"
        class=""
      />
      <div
        aria-hidden="true"
        class=""
      />
      <div
        aria-hidden="true"
        class=""
      />
      <img
        alt="Cleric"
      />
    </div>
    <span>
      Cleric
    </span>
  </button>
  <button
    aria-label="Druid"
    aria-pressed="false"
    class=""
    tabindex="0"
  >
    <div>
      <div
        aria-hidden="true"
        class=""
      />
      <div
        aria-hidden="true"
        class=""
      />
      <div
        aria-hidden="true"
        class=""
      />
      <div
        aria-hidden="true"
        class=""
      />
      <img
        alt="Druid"
      />
    </div>
    <span>
      Druid
    </span>
  </button>
  <button
    aria-label="Sorcerer"
    aria-pressed="false"
    class=""
    tabindex="0"
  >
    <div>
      <div
        aria-hidden="true"
        class=""
      />
      <div
        aria-hidden="true"
        class=""
      />
      <div
        aria-hidden="true"
        class=""
      />
      <div
        aria-hidden="true"
        class=""
      />
      <img
        alt="Sorcerer"
      />
    </div>
    <span>
      Sorcerer
    </span>
  </button>
  <button
    aria-label="Warlock"
    aria-pressed="false"
    class=""
    tabindex="0"
  >
    <div>
      <div
        aria-hidden="true"
        class=""
      />
      <div
        aria-hidden="true"
        class=""
      />
      <div
        aria-hidden="true"
        class=""
      />
      <div
        aria-hidden="true"
        class=""
      />
      <img
        alt="Warlock"
      />
    </div>
    <span>
      Warlock
    </span>
  </button>
  <button
    aria-label="Wizard"
    aria-pressed="false"
    class=""
    tabindex="0"
  >
    <div>
      <div
        aria-hidden="true"
        class=""
      />
      <div
        aria-hidden="true"
        class=""
      />
      <div
        aria-hidden="true"
        class=""
      />
      <div
        aria-hidden="true"
        class=""
      />
      <img
        alt="Wizard"
      />
    </div>
    <span>
      Wizard
    </span>
  </button>
</section>
`;
